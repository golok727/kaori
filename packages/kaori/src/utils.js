export function invariant(condition, message) {
    if (!condition) {
        throw new Error(message);
    }
}
//# sourceMappingURL=utils.js.map