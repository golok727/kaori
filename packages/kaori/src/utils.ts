export function invariant(
  condition: boolean,
  message: string
): asserts condition {
  if (!condition) {
    throw new Error(message);
  }
}

export const has_own = Object.hasOwn;
export const get_own_property_descriptor = Object.getOwnPropertyDescriptor;
export const define_property = Object.defineProperty;
