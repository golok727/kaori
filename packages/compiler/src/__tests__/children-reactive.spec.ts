import { describe, it, expect } from 'vitest';
import { compile } from './utils';

describe('Children prop reactivity', () => {
  it('should make children reactive when containing function calls', async () => {
    const input = `
function Parent() {
  return (
    <Container>
      {getMessage()}
    </Container>
  );
}`;

    const output = await compile(input);
    expect(output).toContain('get children()');
    expect(output).toContain('return getMessage()');
    expect(output).toMatchSnapshot();
  });

  it('should make children reactive when containing member access', async () => {
    const input = `
function Parent() {
  return (
    <Container>
      {user.name}
    </Container>
  );
}`;

    const output = await compile(input);
    expect(output).toContain('get children()');
    expect(output).toContain('user.name');
    expect(output).toMatchSnapshot();
  });

  it('should make children reactive when containing component with function call props', async () => {
    const input = `
function Parent() {
  return (
    <Container>
      <Child title={getTitle()} />
    </Container>
  );
}`;

    const output = await compile(input);
    expect(output).toContain('get children()');
    expect(output).toMatchSnapshot();
  });

  it('should make children reactive when containing html template with member access', async () => {
    const input = `
function Parent() {
  return (
    <Container>
      <p>{user.name}</p>
    </Container>
  );
}`;

    const output = await compile(input);
    expect(output).toContain('get children()');
    expect(output).toMatchSnapshot();
  });

  it('should NOT make children reactive when containing only static content', async () => {
    const input = `
function Parent() {
  return (
    <Container>
      <p>Static text</p>
    </Container>
  );
}`;

    const output = await compile(input);
    expect(output).not.toContain('get children()');
    expect(output).toContain('children: html');
    expect(output).toMatchSnapshot();
  });

  it('should make children reactive even with static component props (component call is dynamic)', async () => {
    const input = `
function Parent() {
  return (
    <Container>
      <Child title="static" />
    </Container>
  );
}`;

    const output = await compile(input);
    expect(output).toContain('get children()');
    expect(output).toContain('component(Child');
    expect(output).toMatchSnapshot();
  });

  it('should make children reactive with mixed static and dynamic content', async () => {
    const input = `
function Parent() {
  return (
    <Container>
      <p>Static</p>
      {getDynamic()}
      <Child prop={getValue()} />
    </Container>
  );
}`;

    const output = await compile(input);
    expect(output).toContain('get children()');
    expect(output).toMatchSnapshot();
  });

  it('should NOT make arrow function children reactive (functions are static)', async () => {
    const input = `
function Parent() {
  return (
    <For items={getItems()}>
      {item => <div>{item.name}</div>}
    </For>
  );
}`;

    const output = await compile(input);
    // Arrow functions are static values - no getter needed
    expect(output).not.toContain('get children()');
    expect(output).toContain('children: item =>');
    expect(output).toContain('item.name');
    expect(output).toMatchSnapshot();
  });

  it('should handle nested reactive children correctly', async () => {
    const input = `
function Parent() {
  return (
    <Outer>
      <Middle>
        <Inner prop={getValue()} />
      </Middle>
    </Outer>
  );
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
  });
});
