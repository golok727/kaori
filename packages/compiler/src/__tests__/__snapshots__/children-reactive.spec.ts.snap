// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Children prop reactivity > should NOT make arrow function children reactive (functions are static) 1`] = `
"import { component, html } from "kaori.js";
function Parent() {
  return html\`\${component(For, {
    get items() {
      return getItems();
    },
    children: item => html\`<div>\${item.name}</div>\`
  })}\`;
}"
`;

exports[`Children prop reactivity > should NOT make children reactive when containing only static content 1`] = `
"import { component, html } from "kaori.js";
function Parent() {
  return html\`\${component(Container, {
    children: html\`<p>Static text</p>\`
  })}\`;
}"
`;

exports[`Children prop reactivity > should handle nested reactive children correctly 1`] = `
"import { component, html } from "kaori.js";
function Parent() {
  return html\`\${component(Outer, {
    get children() {
      return component(Middle, {
        get children() {
          return component(Inner, {
            get prop() {
              return getValue();
            }
          });
        }
      });
    }
  })}\`;
}"
`;

exports[`Children prop reactivity > should make children reactive even with static component props (component call is dynamic) 1`] = `
"import { component, html } from "kaori.js";
function Parent() {
  return html\`\${component(Container, {
    get children() {
      return component(Child, {
        title: "static"
      });
    }
  })}\`;
}"
`;

exports[`Children prop reactivity > should make children reactive when containing component with function call props 1`] = `
"import { component, html } from "kaori.js";
function Parent() {
  return html\`\${component(Container, {
    get children() {
      return component(Child, {
        get title() {
          return getTitle();
        }
      });
    }
  })}\`;
}"
`;

exports[`Children prop reactivity > should make children reactive when containing function calls 1`] = `
"import { component, html } from "kaori.js";
function Parent() {
  return html\`\${component(Container, {
    get children() {
      return getMessage();
    }
  })}\`;
}"
`;

exports[`Children prop reactivity > should make children reactive when containing html template with member access 1`] = `
"import { component, html } from "kaori.js";
function Parent() {
  return html\`\${component(Container, {
    get children() {
      return html\`<p>\${user.name}</p>\`;
    }
  })}\`;
}"
`;

exports[`Children prop reactivity > should make children reactive when containing member access 1`] = `
"import { component, html } from "kaori.js";
function Parent() {
  return html\`\${component(Container, {
    get children() {
      return user.name;
    }
  })}\`;
}"
`;

exports[`Children prop reactivity > should make children reactive with mixed static and dynamic content 1`] = `
"import { component, html } from "kaori.js";
function Parent() {
  return html\`\${component(Container, {
    get children() {
      return [html\`<p>Static</p>\`, getDynamic(), component(Child, {
        get prop() {
          return getValue();
        }
      })];
    }
  })}\`;
}"
`;
